connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java

```diff
@@ -0,0 +1,312 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime.errors;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.RetriableException;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.apache.kafka.common.config.ConfigDef.Range.atLeast;
+import static org.apache.kafka.common.config.ConfigDef.ValidString.in;
+
+/**
+ * <p>
+ * Attempt to recover a failed operation with retries and tolerance limits.
+ * </p>
+ *
+ * <p>
+ * A retry is attempted if the operation throws a {@link RetriableException}. Retries are accompanied by exponential backoffs, starting with
+ * {@link #RETRIES_DELAY_MIN_MS}, up to what is specified with {@link RetryWithToleranceExecutorConfig#retryDelayMax()}.
+ * Including the first attempt and future retries, the total time taken to evaluate the operation should be within
+ * {@link RetryWithToleranceExecutorConfig#retryDelayMax()} millis.
+ * </p>
+ *
+ * <p>
+ * This executor will tolerate failures, as specified by {@link RetryWithToleranceExecutorConfig#toleranceLimit()}.
+ * For transformations and converters, all exceptions are tolerated. For others operations, only {@link RetriableException} are tolerated.
+ * </p>
+ *
+ * <p>
+ * There are three outcomes to executing an operation. It might succeed, in which case the result is returned to the caller.
+ * If it fails, this class does one of these two things: (1) if the failure occurred due to a tolerable exception, then
+ * set appropriate error reason in the {@link ProcessingContext} and return null, or (2) if the exception is not tolerated,
+ * then it is wrapped into a ConnectException and rethrown to the caller.
+ * </p>
+ */
+public class RetryWithToleranceExecutor {
```

Comment:
Glad to see `OperationExecutor` interface was removed. Haven't read all the comments and I'm not sure what was the reason. But I find the `Executor` suffix misleading if it doesn't implement the `Executor` interface (which is a bit specific in the java library). Should we consider a removal from here too? Not too strong about this and the problem is that the executor name is everywhere in the code (but not in the keep). Still a bit misleading because this class is not an executor in the narrow sense.

Reply:
Any thoughts on what else this class could be called? I understand the conflict with java's Executor? But other words (for example, evaluator, handler) didn't really fit the bill too. In a sense this class executes `Operations`, whereas Java's executors execute `Runnable`. In that sense, there is some similarities. Since this name was chosen when we had an interface, should we simply call this class `OperationExecutor`?

Reply:
`RetryWithToleranceOperator`. Basically, an operation executor is ... an operator. The word has somehow specific meaning too, but in a different context, so maybe it can't be confused that easily. 

Using Operator instead of Executor could leave most of your javadocs unchanged and could also have an `execute` method (not too bad). Just a suggestion.

Reply:
Renamed the class. I suppose the biggest challenge to calling it an operator was that usually multiple operators typically exist/work together (arithmetic or query operators, for example). It's a bit weird to have a single operator. But, as you said, it does reduce confusion.